ParameterState: Add optional IEqualityComparer

ScarletKuro opened the pull request:
## Description
Fixes when custom comparison is required.
![IEqualityComparer](https://github.com/MudBlazor/MudBlazor/assets/19953225/0c600b0e-2b6a-40e0-9f75-61990cea671a)

### Replaced non generic `ParameterState` with a builder.
Builder pattern is really good when overloads grow exponentially, so I had to drop the non generic `ParameterState` helper for the builder pattern. While it's still manageable for `RegisterParameters` it wasn't  the case for the `ParameterState` helper.
It's much easier to use now `ParameterAttachBuilder`, it has own drawbacks like you need to set the `T` explicitly, it can't get it from the context, BUT `ParameterAttachBuilder` **should not** be used by anyone, not by users and not by contributors, it's exclusively for the tests and the `RegisterParameters` to build complex `ParameterState<T>` which has so much optional arguments and in future it may only grow.

## How Has This Been Tested?
None as for now, unit tests are work in progress

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [ ] I've added relevant tests.

ScarletKuro committed:
wip

ScarletKuro committed:
wip

ScarletKuro committed:
Add Builder and drop non generic ParameterSet

henon reviewed (approved):
The builder makes a lot of sense. I didn't know it at the time but the gut decision for the `RegisterParameter` method in the component constructor was the right one, seeing that it is now able to abstract a lot of complexity away from the framework user.

After reading our CONTRIBUTING.md, people still seem to be managing to shoot themselves in the foot with the framework. I will think a bit on how to improve the docs without making them overly long and exhausting to read.

ScarletKuro committed:
Add ParameterViewExtensionsTests

ScarletKuro committed:
Add more tests

ScarletKuro commented:
In this PR I also moved the `NumericConverter.AreEqual` logic to the separate `DoubleEpsilonEqualityComparer`, because:
* It how it should be from the beginning
* I used the `DoubleEpsilonEqualityComparer` for the `SetParametersAsync_CustomComparer_HandlerShouldFire` and `SetParametersAsync_CustomComparer_HandlerShouldNotFire` tests.

I also added tests for `DoubleEpsilonEqualityComparer`, since this logic is a copy-paste from the java implementation https://floating-point-gui.de/errors/comparison/ I also copied the tests from https://floating-point-gui.de/errors/NearlyEqualsTest.java but not all of them, because we use different `MinNormal` compared to the JAVA version, and the tests are specifically optimized for the `FloatMinNormal = (1 << 23) * float.Epsilon`;

ScarletKuro committed:
Cleanup

ScarletKuro committed:
Add more tests.

ScarletKuro committed:
Add bUnit test

ScarletKuro commented:
Added bUnit test. Merging as the feature is complete and fully tested.

ScarletKuro commented:
@henon now when I look at some components I think I made a mistake and the `IEqualityComparer` should be a `Func` because some components have `IEqualityComparer` defined as [Parameter] and it's used dynamically for specific parameterю
For examle in tree the
``
[Parameter]
[Category(CategoryTypes.TreeView.Selecting)]
public IEqualityComparer<T?> Comparer { get; set; } = EqualityComparer<T?>.Default;
```
And it's used for the `SelectedValue` and this is a problem since `Comparer` can change at any time, while currently we set in ctor.
We could just create `SetComparer` in `IParameterState`, but then you'd have to change it yourself everytime, therefore it would be nice if it would be automatic via `Func`.

henon commented:
> ... therefore it would be nice if it would be automatic via `Func`.

Agreed
